pallene/to_ir.lua

```diff
@@ -38,15 +38,26 @@ function ToIR:init()
     self.fun_id_of_exp  = {} -- { ast.Exp  => integer }
     self.glb_id_of_decl = {} -- { ast.Decl => integer } -- non-exported globals
     self.glb_id_of_var  = {} -- { ast.Var  => integer } -- exported global
+    self.func_stack     = {} -- list of function metadata
 end
 
 function ToIR:enter_function(f_id)
     -- Function-specific variables
     -- These are re-initialized each time.
-    self.func = self.module.functions[f_id]
-    self.loc_id_of_decl = {} -- { ast.Decl => integer }
-    self.call_exps      = {} -- { ast.Exp.CallFunc }
-    self.dsts_of_call   = {} -- { ast.Exp => { var_id } }
+    local func_data = {
+        loc_id_of_decl = {}, -- { ast.Decl => integer }
+        call_exps      = {}, -- { ast.Exp.CallFunc }
+        dsts_of_call   = {}, -- { ast.Exp => { var_id } }
+        func           = self.module.functions[f_id],
+        f_id           = f_id
+    }
+
+    table.insert(self.func_stack, func_data)
+
+    self.func           = func_data.func
+    self.loc_id_of_decl = func_data.loc_id_of_decl
+    self.call_exps      = func_data.call_exps
+    self.dsts_of_call   = func_data.dsts_of_call
```

Comment:
In theory, I think the call_exps and the dsts_of_call would still work if they were part of the `self` instead of part of the function metadata. In other words, I think they would also work if they were like the `fun_id_of_decl`, which is not reset when we enter and exit functions.

I don't know if I prefer having them inside the function metadata or outside the function metadata. Which one do you think is better?

Reply:
For me, I think that we need a stack to keep track of the relevant metadata anyway.
I prefer this since it's nicer to *my* eyes if all the bookkeeping information about a function is kept in the same place.
That said, I'm not sure if this is the way to go.

Reply:
The reason I'm asking this is because the `call_exps` and the `dsts_of_call` are not about the current function that we are in. They're about the function *calls* that are present in the program.